### PR TITLE
Will's Onboarding Task

### PR DESCRIPTION
**Pull request by Will-est**
# feature/Will

PR Link: https://github.com/VEXU-GHOST/VEXU_GHOST/pull/35#issue-2045397549

Issue Link: N/A

### Description
Merges my software onboarding 2 to onboarding.

### Reviewers

Required: @Maxx

---
### Changelog

- Made node that outputs the result of a mathematical operation to a topic called "/output_topic" based on an input to a topic called "/input_topic" 

- Made formatted to cpp/hpp format

- added Gtest

- finalized package

### Reviewer Guide

- Unsure about my linting or how uncrustified my code is

#### Automatic
- Really just tested my decode function with all 4 possibilites(adding, multiplying, subtracting, and modulusing) and a random input
#### Manual
- Basically just ran the node and fed different inputs to it, all of which worked.

### Documentation
N/A

### Checklist
- [x] Confirmed all tests pass on a clean build
- [x] Added reviewers in Github
- [x] Posted PR Summary to Discord PR's Channel
- [x] Ran uncrustify on any modified C++ files
- [x] Ran Colcon Lint for any modified CMakeLists.txt or Package.xml

### CODE REVIEW

File: 10_Examples/will_onboarding1/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  set(ament_cmake_copyright_FOUND TRUE)
+  # the following line skips cpplint (only works in a git repo)
+  # comment the line when this package is in a git repo and when
+  # a copyright and license is added to all source files
+  set(ament_cmake_cpplint_FOUND TRUE)
+  ament_lint_auto_find_test_dependencies()
+endif()
+
+#Creating a library for the class that I made
+add_library(Math_Node SHARED src/Mather.cpp)
```
Comment (MaxxWilson) [inline 10_Examples/will_onboarding1/CMakeLists.txt]: either pick CamelCase or snake_case, don't combine both lol.

For cmake, I believe we do snake_case.

Comment (MaxxWilson) [reply 10_Examples/will_onboarding1/CMakeLists.txt]: This applies to whole file

File: 10_Examples/will_onboarding1/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  TARGETS Math_Node
+  EXPORT Math_Node
+  LIBRARY DESTINATION lib
+  ARCHIVE DESTINATION lib
+  RUNTIME DESTINATION bin
+  INCLUDES DESTINATION include
+)
+
+#Main Method Stuff
+add_executable(main_executable src/main.cpp)
```
Comment (MaxxWilson) [inline 10_Examples/will_onboarding1/CMakeLists.txt]: More descriptive executable name? "MathNodeMain"

File: 10_Examples/will_onboarding1/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+target_link_libraries(main_executable
+Math_Node)
+
+install(TARGETS
+main_executable
+DESTINATION lib/${PROJECT}
+)
+
+#Gtest stuff
+ament_add_gtest(test_example src/wills_Test_File.cpp)
```
Comment (MaxxWilson) [inline 10_Examples/will_onboarding1/CMakeLists.txt]: CamelCase snake_case thing again lol.

File: 10_Examples/will_onboarding1/include/will_onboarding1/Mather.hpp [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+#include <memory>
+#include <string>
+
+#include "rclcpp/rclcpp.hpp"
+#include "std_msgs/msg/int32.hpp"
+
+using namespace std::chrono_literals;
+using namespace std;
+using std::placeholders::_1;
+
```
Comment (MaxxWilson) [inline 10_Examples/will_onboarding1/include/will_onboarding1/Mather.hpp:11]: "using" is bad practice in a header file (some instances I may be guilty).

"using namespace" is REALLY bad practice.

Anything that includes this header (and any headers which include a header which include this) will get all of these propogated and basically removes namespacing all together, i.e. big trouble.

File: 10_Examples/will_onboarding1/include/will_onboarding1/Mather.hpp [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+namespace Math_Space {
```
Comment (MaxxWilson) [inline 10_Examples/will_onboarding1/include/will_onboarding1/Mather.hpp]: we do snake case for namespaces, i.e. math_space.

Also putting space at the end of namespace is a bit odd.

File: 10_Examples/will_onboarding1/include/will_onboarding1/Mather.hpp [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+class Mather : public rclcpp::Node {
+public:
+	Mather(int a, int b);
+
+	int decode(int encoded);
+
+	void topic_callback(const std_msgs::msg::Int32::SharedPtr msg);
+
+	void timer_callback();
```
Comment (MaxxWilson) [inline 10_Examples/will_onboarding1/include/will_onboarding1/Mather.hpp:49]: Header files are sometimes the only thing people can count on when using an external library, so this is where you would put comment blocks above every method describing what it does, if it throws errors, what edge cases, etc.

Example:
image

File: 10_Examples/will_onboarding1/include/will_onboarding1/Mather.hpp [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+private:
+	int a;
+	int b;
+	int decodeInt;
```
Comment (MaxxWilson) [inline 10_Examples/will_onboarding1/include/will_onboarding1/Mather.hpp]: member variables are always named either m_thing, or thing_. 

Which one depends on who you are working with. I do m_ these days, but may have been trailing underscore when some of GHOST code was written.

**Comment by MaxxWilson** [comment]
Overall, pretty solid.

Mostly small (but important things). Only high-level issue I'd point out is that usually library should be separated from ROS Node so you can test them separately. So in this case, Math would be its own library and then youd have a separate node which uses it.

That removes ROS dependency from your test and then people can use your code without requiring ROS.

File: 10_Examples/will_onboarding1/include/will_onboarding1/Mather.hpp (hunk 5 above)
Comment (Will-est) [reply 10_Examples/will_onboarding1/include/will_onboarding1/Mather.hpp]: the space after the namespace comes from the auto formatter.
